x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -520,152 +534,218 @@ public void invalidateRefreshToken(String refreshToken, ActionListener<Boolean>
-                                    ActionListener<Boolean> wrappedListener =
-                                            ActionListener.wrap(ignore -> listener.onResponse(false), listener::onFailure);
-                                    indexInvalidation(tokenDocId, version, wrappedListener, attemptCount, "access_token", 1L);
-                                } else if (isShardNotAvailableException(e)) {
-                                    attemptCount.incrementAndGet();
-                                    indexBwcInvalidation(userToken, listener, attemptCount, expirationEpochMilli);
-                                } else {
-                                    listener.onFailure(e);
-                                }
-                            }), client::index));
+                bulkRequestBuilder.add(indexRequest);
+            }
+            bulkRequestBuilder.setRefreshPolicy(RefreshPolicy.WAIT_UNTIL);
+            final BulkRequest bulkRequest = bulkRequestBuilder.request();
+            securityIndex.prepareIndexIfNeededThenExecute(ex -> listener.onFailure(traceLog("prepare security index", ex)),
+                () -> executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, bulkRequest,
+                ActionListener.<BulkResponse>wrap(bulkResponse -> {
+                    List<String> retryTokenIds = new ArrayList<>();
+                    for (BulkItemResponse bulkItemResponse : bulkResponse.getItems()) {
+                        if (bulkItemResponse.isFailed()) {
+                            Throwable cause = bulkItemResponse.getFailure().getCause();
+                            logger.error(cause.getMessage());
+                            traceLog("(bwc) invalidate tokens",  cause);
+                            if (isShardNotAvailableException(cause)) {
+                                retryTokenIds.add(getTokenIdFromInvalidatedTokenDocumentId(bulkItemResponse.getFailure().getId()));
+                            } else if ((cause instanceof VersionConflictEngineException) == false){
+                                // We don't handle VersionConflictEngineException, the ticket has been invalidated
+                                listener.onFailure(bulkItemResponse.getFailure().getCause());
+                            }
+                        }
+                    }
+                    if (retryTokenIds.isEmpty() == false) {
+                        attemptCount.incrementAndGet();
+                        indexBwcInvalidation(retryTokenIds, listener, attemptCount, expirationEpochMilli, previousResult);
+                    }
+                    indexInvalidation(tokenIds, listener, attemptCount, "access_token", previousResult);
+                }, e -> {
+                    Throwable cause = ExceptionsHelper.unwrapCause(e);
+                    traceLog("(bwc) invalidate tokens", cause);
+                    if (isShardNotAvailableException(e)) {
```

Comment:
hmm I wonder if this should be cause instead of e?

Reply:
Good catch!